gateway/ln-gateway/src/gateway_module_ng/receive_sm.rs

```diff
@@ -0,0 +1,243 @@
+                        move |_, error, old_state| {
+                            Box::pin(Self::transition_funding_rejected(error, old_state))
+                        },
+                    ),
+                    StateTransition::new(
+                        Self::await_outcome_ready(
+                            global_context.clone(),
+                            context.decoder.clone(),
+                            context.tpe_pks.clone(),
+                            self.out_point,
+                            self.contract.clone(),
+                        ),
+                        move |dbtx, output_outcomes, old_state| {
+                            Box::pin(Self::transition_outcome_ready(
+                                dbtx,
+                                output_outcomes,
+                                old_state,
+                                gc.clone(),
+                                tpe_agg_pk,
+                            ))
+                        },
+                    ),
+                ]
+            }
+            ReceiveSMState::Success(..) => {
+                vec![]
+            }
+            ReceiveSMState::Rejected(..) => {
+                vec![]
+            }
+            ReceiveSMState::Refunding(..) => {
+                vec![]
+            }
+            ReceiveSMState::Failure => {
+                vec![]
+            }
+        }
+    }
+
+    fn operation_id(&self) -> OperationId {
+        self.operation_id
+    }
+}
+
+impl ReceiveStateMachine {
+    async fn await_funding_rejected(
+        global_context: DynGlobalClientContext,
+        txid: TransactionId,
+    ) -> String {
+        match global_context.await_tx_accepted(txid).await {
+            Ok(()) => pending::<String>().await,
+            Err(error) => error,
+        }
+    }
+
+    async fn transition_funding_rejected(
+        error: String,
+        old_state: ReceiveStateMachine,
+    ) -> ReceiveStateMachine {
+        ReceiveStateMachine {
+            operation_id: old_state.operation_id,
+            contract: old_state.contract,
+            out_point: old_state.out_point,
+            state: ReceiveSMState::Rejected(error),
+        }
+    }
+
+    async fn await_outcome_ready(
+        global_context: DynGlobalClientContext,
+        module_decoder: Decoder,
+        tpe_pks: BTreeMap<PeerId, PublicKeyShare>,
+        out_point: OutPoint,
+        decryption_contract: IncomingContract,
+    ) -> BTreeMap<PeerId, DecryptionKeyShare> {
+        let verify_decryption_share = move |peer, outcome: SerdeOutputOutcome| {
+            let outcome = deserialize_outcome::<LightningOutputOutcome>(outcome, &module_decoder)?;
+
+            match outcome {
+                LightningOutputOutcome::Incoming(share) => {
+                    if !decryption_contract.verify_decryption_share(
+                        tpe_pks.get(&peer).ok_or(anyhow!("Unknown peer pk"))?,
```

Comment:
We probably will want to make these error messages (here and below) slightly more descriptive in case this ever happens in a production deployment